core/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java

```diff
@@ -148,47 +155,84 @@ Collector create(Collector in) throws IOException {
+        void postProcess(QuerySearchResult result) throws IOException {
             result.topDocs(topDocsCollector.getTopDocs(), sortFmt);
         }
     }
 
     abstract static class SimpleTopDocsCollectorContext extends TopDocsCollectorContext {
         private final @Nullable SortAndFormats sortAndFormats;
-        private final TopDocsCollector<?> topDocsCollector;
+        private final Collector collector;
+        private final IntSupplier totalHitsSupplier;
+        private final Supplier<TopDocs> topDocsSupplier;
 
         /**
          * Ctr
+         * @param reader The index reader
+         * @param query The Lucene query
          * @param sortAndFormats The query sort
          * @param numHits The number of top hits to retrieve
          * @param searchAfter The doc this request should "search after"
          * @param trackMaxScore True if max score should be tracked
+         * @param trackTotalHits True if the total number of hits should be tracked
+         * @param hasFilterCollector True if the collector chain contains at least one collector that can filters document
          */
-        private SimpleTopDocsCollectorContext(@Nullable SortAndFormats sortAndFormats,
+        private SimpleTopDocsCollectorContext(IndexReader reader,
+                                              Query query,
+                                              @Nullable SortAndFormats sortAndFormats,
                                               @Nullable ScoreDoc searchAfter,
                                               int numHits,
-                                              boolean trackMaxScore) throws IOException {
+                                              boolean trackMaxScore,
+                                              boolean trackTotalHits,
+                                              boolean hasFilterCollector) throws IOException {
             super(REASON_SEARCH_TOP_HITS, numHits);
             this.sortAndFormats = sortAndFormats;
             if (sortAndFormats == null) {
-                this.topDocsCollector = TopScoreDocCollector.create(numHits, searchAfter);
+                final TopDocsCollector<?> topDocsCollector = TopScoreDocCollector.create(numHits, searchAfter);
+                this.collector = topDocsCollector;
+                this.topDocsSupplier = topDocsCollector::topDocs;
+                this.totalHitsSupplier = topDocsCollector::getTotalHits;
             } else {
-                this.topDocsCollector = TopFieldCollector.create(sortAndFormats.sort, numHits,
-                    (FieldDoc) searchAfter, true, trackMaxScore, trackMaxScore, true);
+                final boolean canEarlyTerminate = canEarlyTerminate(reader, sortAndFormats);
+                final TopDocsCollector<?> topDocsCollector = TopFieldCollector.create(sortAndFormats.sort, numHits,
+                    (FieldDoc) searchAfter, true, trackMaxScore, trackMaxScore, canEarlyTerminate == false);
```

Comment:
I could be wrong, but I think we could simplify it a bit by doing something like that:

```
// implicit total hit counts are valid only when there is no filter collector in the chain
int count = hasFilterCollector ? -1 : shortcutTotalHitCount(reader, query);
boolean doTrackTotalHits = trackTotalHits && count == -1; // we can also skip total hit counts if the query gives it for free
final TopDocsCollector<?> topDocsCollector = TopFieldCollector.create(sortAndFormats.sort, numHits,
    (FieldDoc) searchAfter, true, trackMaxScore, trackMaxScore, doTrackTotalHits);
```

This way, we don't need to check whether the collector can actually early-terminate or not, and we never have to add a TotalHitCount collector?

Reply:
We still want the top docs collector to early terminate if `track_total_hits` is true. 
This is why we have the complex logic that wraps a counting collector in the next block.
We could also add a filtered collector that intercepts `CollectionTerminatedException` coming from the top docs collector and continue the collection with a simple counting collector to get the total hit count. This way we can always assume that the TopDocsCollector can early terminate ?

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/27666/commits/09a033aee231f90de57153e19283ed496e3a3b46 to simplify the logic. Can you take another look ?